### DISCUSSION TITLE
.vox -> obj -> glb with gltfpack renders corrert with babylon but not with three

### DISCUSSION THREAD

**Discussion by arpu** [zeux/meshoptimizer#269] [Q&A] [answered]
HI,

not sure this is a gltfpack problem or threejs 

zip obj mtl + png and result glb 
https://drive.google.com/file/d/1wCXw9GYepyUm9DZt_gFDaZaBWx2qpBHL/view?usp=sharing

gltfpack output

```
gltfpack -tp -tc -i 1-Monkey.obj -o test_monkey.glb -vv
toktx: error while loading shared libraries: libktx.so.4: cannot open shared object file: No such file or directory
toktx --version => 32512
Basis Universal GPU Texture Compressor Reference Encoder v1.12.00
Copyright (C) 2019-2020 Binomial LLC, All rights reserved
basisu -version => 0
input: 0 nodes, 0 meshes (1 primitives), 1 materials, 0 skins, 0 animations
input: 1 mesh primitives (1064 triangles, 3192 vertices); 1 draw calls (1 instances, 1064 triangles)
basisu -mipmap -comp_level 1 -q 192 -file /tmp/gltfpack-XhqDWo.png -output_file /tmp/gltfpack-C6mMDp.basis => 0
output: 1 mesh primitives (1064 triangles, 1606 vertices); 1 draw calls (1 instances, 1064 triangles)
output: 1 nodes, 1 meshes (1 primitives), 1 materials
output: JSON 1539 bytes, buffers 33436 bytes
output: buffers: vertex 25696 bytes, index 6384 bytes, skin 0 bytes, time 0 bytes, keyframe 0 bytes, instance 0 bytes, image 1355 bytes
stats: vertex POSITION: compressed 12848 bytes (64.0 bits), raw 12848 bytes (64 bits)
stats: vertex NORMAL: compressed 6424 bytes (32.0 bits), raw 6424 bytes (32 bits)
stats: vertex TEXCOORD: compressed 6424 bytes (32.0 bits), raw 6424 bytes (32 bits)
stats: index index: compressed 6384 bytes (16.0 bits), raw 6384 bytes (16 bits)

```

render in three viewer https://gltf-viewer.donmccurdy.com/

![Bildschirmfoto von 2021-04-11 14-21-48](https://user-images.githubusercontent.com/14173/114304019-49af4e00-9ad1-11eb-8211-67e2cad0930f.png)

babylon sandbox

![Bildschirmfoto von 2021-04-11 14-22-36](https://user-images.githubusercontent.com/14173/114304044-63e92c00-9ad1-11eb-82cc-860f9e3c89ba.png)

the obj  exported in MagicaVoxel-0.99.6.3-win64 from .vox file

**Comment by arpu** [comment]
wups i see the error :> toktx: error while loading shared libraries: libktx.so.4: cannot open shared object file: No such file or directory

**Comment by arpu** [comment, marked as answer]
ok updated to lates KTX_Software

glb in viewer looks now
![Bildschirmfoto von 2021-04-11 14-29-10](https://user-images.githubusercontent.com/14173/114304214-4cf70980-9ad2-11eb-914a-dc644081fd5d.png)

```
gltfpack -tp -tc -i 1-Monkey.obj -o test_monkey.glb -vv
toktx v4.0.0-rc1~6
toktx --version => 0
input: 0 nodes, 0 meshes (1 primitives), 1 materials, 0 skins, 0 animations
input: 1 mesh primitives (1064 triangles, 3192 vertices); 1 draw calls (1 instances, 1064 triangles)
toktx --t2 --2d --genmipmap --nowarn --resize 256x4 --bcmp --clevel 1 --qlevel 192 --srgb -- /tmp/gltfpack-3kluxB.ktx2 /tmp/gltfpack-hxGzrD.png => 0
output: 1 mesh primitives (1064 triangles, 1606 vertices); 1 draw calls (1 instances, 1064 triangles)
output: 1 nodes, 1 meshes (1 primitives), 1 materials
output: JSON 1539 bytes, buffers 33188 bytes
output: buffers: vertex 25696 bytes, index 6384 bytes, skin 0 bytes, time 0 bytes, keyframe 0 bytes, instance 0 bytes, image 1107 bytes
stats: vertex POSITION: compressed 12848 bytes (64.0 bits), raw 12848 bytes (64 bits)
stats: vertex NORMAL: compressed 6424 bytes (32.0 bits), raw 6424 bytes (32 bits)
stats: vertex TEXCOORD: compressed 6424 bytes (32.0 bits), raw 6424 bytes (32 bits)
stats: index index: compressed 6384 bytes (16.0 bits), raw 6384 bytes (16 bits)
```

**Reply by zeux** [reply]
This is what I get when using toktx with https://gltf-viewer.donmccurdy.com/

![image](https://user-images.githubusercontent.com/1106629/114331641-4def6180-9af9-11eb-8dec-9c94ad5a1ecf.png)

Unsure why this is different from your output. I'm using `toktx v4.0.beta1.380.g0d851050`; I'll try the latest rc.

When you use basisu (your first attempt), three.js fails to decode the resulting image for some reason with "[.WebGL-00001C6C0015BA80] GL_INVALID_OPERATION: Invalid compressed image size." in the error log. I'm not sure why that is, will need to debug - could be that KTX2 spec changed again, could be something specific to the image size (256x1).

**Reply by zeux** [reply]
Ah, right, basisu doesn't support texture resizing. The image is 256x1 but WebGL mandates that the image must be divisible by 4 for compressed inputs. gltfpack automatically fixes this problem when toktx is used, but not when basisu is used.

When using toktx 4.0-rc1, I get this in three.js:

![image](https://user-images.githubusercontent.com/1106629/114332881-ee468580-9afb-11eb-90ea-c36c1a89acfa.png)

Which looks correct to me, this is what it looks like without using -tc:

![image](https://user-images.githubusercontent.com/1106629/114332960-13d38f00-9afc-11eb-9be5-83093be246a0.png)

So I'm still not sure why you're getting a different result from me.

**Reply by arpu** [reply]
thx for looking at this, 

some findings: 
without -tc  the glb renders fine on Linux chome/firefox and threejs viewer https://gltf-viewer.donmccurdy.com/ ( r126) 

with -tc and -tp 
(toktx --t2 --2d --genmipmap --nowarn --resize 256x4 --bcmp --clevel 1 --qlevel 192 --srgb -- /tmp/gltfpack-wGbFFB.ktx2 /tmp/gltfpack-ThRHCC.png => 0
)

firefox/chrome on linux no colors 

will test the glb on windows later

**Reply by arpu** [reply]
ok looks like the problem comes from newer master KTX-Software

version i use is from master aee7a1c5d01c97f22d8a931f7c3f09c2e2e87a7f

**Reply by arpu** [reply]
i can confirm with KTX-Software  toktx v4.0.0-beta8~5 7ad9caf1efd4d970fb993867dbe4c720552e6b02

the ktx2 glb verson works fine in threejs and babylon on Linux

**Reply by zeux** [reply]
I recommend comparing the output of `ktxinfo 1-Monkey.ktx2` after running the following command with both versions of toktx:

```
toktx --t2 --2d --genmipmap --nowarn --resize 256x4 --bcmp --clevel 1 --qlevel 192 --srgb 1-Monkey.ktx2 1-Monkey.png 
```

I suspect what happens is that after https://github.com/KhronosGroup/KTX-Software/commit/2ffdc81a5c985cfa81d6e216e34676f59c58468a the color primaries information in the output file is wrong, and on Linux it results in transcoding to a luminance-alpha format mistakenly (whereas on Windows it's not available so it works fine)

*edit* although maybe it's not that, I ran older & newer toktx myself and the DFD is the same between these. Not sure what exactly the problem is then.

**Reply by arpu** [reply]
ok not sure what it fixed but with latest KTX_Software 3acee50ae009f5d5b4e4dbacbc3a515d4a7b89ef all looks fine now! 

i think this Question is Answered.
